hadml/datamodules/gan_datamodule.py

```diff
@@ -251,3 +251,196 @@ def test_dataloader(self):
+
+        print('\n', ' '*20, self.__class__.__name__,'\n', '-'*70, sep='')
+
+        if initialise_data_preparation:
+            self.prepare_data()
+            self.setup()
+            one_train_batch = next(iter(self.train_dataloader()))
+            gen_input, disc_input = one_train_batch
+            print(f"\nGenerator input batch: {len(gen_input)} ([{gen_input.size()}])")
+            print(f"Discriminator input batch: {len(disc_input)} ([{disc_input.size()}])")
+            print("\nGenerator input sample", gen_input[0])
+            print("\nDiscriminator input sample", disc_input[0])
+
+    def prepare_data(self):
+        # Load data prepared by the parser
+        with open(self.processed_filename, "rb") as f:
+            data = np.load(f, allow_pickle=True)
+        
+        cluster_kin = data.item()["cluster_kin"]
+        hadron_kin = data.item()["had_kin"]
+        had_type_indices = data.item()["had_type_indices"]
+        cluster_labels = data.item()["cluster_labels"]
+        n_events = len(cluster_kin)
+        self.n_had_types = data.item()["n_had_type_indices"] + 1 # 1 extra type for a stop/padding token
+
+        # Preparing distribution plots
+        distplots_path = os.path.join(os.path.normpath(self.data_dir), 
+                                      os.path.normpath("processed/distribution_plots.pdf"))
+        if not os.path.exists(distplots_path):
+            n_hadrons = [len(d) for d in hadron_kin]
+            hadron_energy = np.concatenate([d for d in hadron_kin])[:, 0]
+            cluster_energy = np.concatenate([d for d in cluster_kin])[:, 0]
+            self.__plot_dist(distplots_path, data=[n_hadrons, hadron_energy, cluster_energy],
+                xlabels=["Number of hadrons", "Energy [GeV]", "Energy [GeV]"],
+                ylabels=["Number of clusters", "Number of hadrons", "Number of clusters"],
+                legend_labels=["Hadron Distribution", "Hadron Energy Distribution",
+                            "Cluster Energy Distribution"])
+
+        # Changing hadron kinematics to the values computed for the cluster rest frame
+        hadron_kin = data.item()["had_kin_rest_frame"]
+
+        # Assigning hadrons to clusters 
+        self.clusters, self.hadrons_with_types, n = self.__get_hadrons_and_clusters__(
+            n_events, cluster_kin, cluster_labels, hadron_kin, had_type_indices)
+        n_clusters_extracted_from_events = n
+
+        self.max_n_hadrons = max([len(hadron_with_type[0]) for hadron_with_type in 
+                                  self.hadrons_with_types])
+        
+        print("Initial number of events:", n_events)
+        print("Total number of clusters:", n_clusters_extracted_from_events)
+        print("Total number of hadron types (with a stop token):", self.n_had_types)
+        print("Largest number of hadrons per cluster:", self.max_n_hadrons)
+
+    def __get_hadrons_and_clusters__(self, n_events, cluster_kin, cluster_labels, hadron_kin, 
```

Comment:
Double leading/trailing underscores are generally only used for python "magic" methods. Here, one leading underscore should be enough: `_get_hadrons_and_clusters`

see: https://peps.python.org/pep-0008/#method-names-and-instance-variables